aws/resource_aws_iam_service_linked_role.go

```diff
@@ -0,0 +1,169 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsIamServiceLinkedRole() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsIamServiceLinkedRoleCreate,
+		Read:   resourceAwsIamServiceLinkedRoleRead,
+		Delete: resourceAwsIamServiceLinkedRoleDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"aws_service_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, es []error) {
+					value := v.(string)
+					if !strings.HasSuffix(value, ".amazonaws.com") {
+						es = append(es, fmt.Errorf(
+							"%q must be a service URL e.g. elasticbeanstalk.amazonaws.com", k))
+					}
+					return
+				},
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"path": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"create_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"unique_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsIamServiceLinkedRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iamconn
+
+	serviceName := d.Get("aws_service_name").(string)
+
+	params := &iam.CreateServiceLinkedRoleInput{
+		AWSServiceName: aws.String(serviceName),
+	}
+
+	_, err := conn.CreateServiceLinkedRole(params)
+
+	if err != nil {
+		return fmt.Errorf("Error creating service-linked role with name %s", serviceName)
+	}
+
+	d.SetId(serviceName)
+
+	return nil
+}
+
+func resourceAwsIamServiceLinkedRoleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iamconn
+
+	params := &iam.ListRolesInput{
+		PathPrefix: aws.String(fmt.Sprintf("/aws-service-role/%s/", d.Id())),
+	}
+
+	resp, err := conn.ListRoles(params)
+
+	if err != nil {
+		return err
+	}
+
+	if resp == nil || len(resp.Roles) == 0 {
+		d.SetId("")
+		return nil
+	}
+
+	role := resp.Roles[0]
+
+	d.Set("name", role.RoleName)
+	d.Set("path", role.Path)
+	d.Set("arn", role.Arn)
+	d.Set("create_date", role.CreateDate)
+	d.Set("unique_id", role.RoleId)
+	d.Set("description", role.Description)
+
```

Comment:
When we update `d.SetId()` to the role ARN, we should call `d.Set("aws_service_name", ...)` where we determine the service name from the ARN. This is to prevent the difference on import which will mark the resource as needing recreation.